lib/cli/drive-selection-widget.js

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2016 resin.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const chalk = require('chalk');
+const inquirer = require('inquirer');
+const _ = require('lodash');
+
+const DynamicList = require('./dynamic-list');
+const scanner = require('./drive-scanner');
+const driveComparer = require('../shared/utils').driveComparer;
+const constraints = require('../shared/drive-constraints');
+
+inquirer.registerPrompt('dynamicList', DynamicList);
+
+const driveToChoiceFn = (image) => {
+
+  /**
+   * Map a drive object to a `inquirer` choice object
+   * @param {object} drive - Drive object as returned by `drivelist`
+   * @private
+   * @returns {{name: string, value: object}}
+   */
+  const driveToChoice = (drive) => {
+    const size = drive.size / 1000000000;
+
+    const driveInfo = `${drive.device} (${size.toFixed(1)} GB) - ${drive.description}`;
+
+    const systemDriveText = chalk.red(constraints.labels.SYSTEM_DRIVE);
+    const systemDriveDisplay = constraints.isSystemDrive(drive) ? systemDriveText : '';
+
+    const notRecommendedSizeText = chalk.yellow(constraints.labels.NOT_RECOMMENDED);
+    const recommendedSizeDisplay = constraints.isDriveSizeRecommended(drive, image) ? '' : notRecommendedSizeText;
+
+    const displayName = `${driveInfo} ${systemDriveDisplay} ${recommendedSizeDisplay}`;
+
+    const choice = {
+      name: displayName,
+      value: drive
+    };
+
+    if (!constraints.isDriveLargeEnough(drive, image)) {
+      choice.disabled = constraints.labels.TOO_SMALL;
```

Comment:
Oh, I see. So its some kind of informative message you show to explain why an element is disabled, but it doesn't control whether the option is disabled or not, right?

If so, I think it might be good to add a comment explaining it, since it looks really awkward at first sight (definitely not your fault. The interface exposed by inquirer for this is very weird).